### ISSUE TITLE
Docker GUI does not start properly with specific values in the no_proxy environment variable.

### ISSUE THREAD

**Issue by cyclump** [docker/for-win#5551]
<!--
Please, check https://docs.docker.com/docker-for-windows/troubleshoot/.
Issues without logs and details cannot be debugged, and will be closed.

Issues unrelated to Docker Desktop will be closed.  In particular, see
  - https://github.com/moby/moby/issues for Docker daemon, e.g. running on Windows Server with Docker EE
  - https://github.com/docker/compose/issues for docker-compose
  - https://github.com/docker/machine/issues for docker-machine
  - https://github.com/docker/docker.github.io/issues for the documentation
-->

<!-- Click these checkboxes after submitting. -->
<!-- Download Docker Desktop 'Edge' (latest build) here: https://hub.docker.com/editions/community/docker-ce-desktop-windows -->
  - [ ] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Expected behavior
Docker Desktop for Windows should show the settings in the settings window after right clicking the icon in the task bar and selecting Settings.

### Actual behavior
We are behind a man in the middle type corporate proxy.  We use the http_proxy, https_proxy, and no_proxy environment variables to control how CLI apps access the proxy.  The no_proxy environment variable contains entries like .my-network.net and my-network.net.  Some tools need a leading dot and some don't.  All of our tools don't fail with both in the no_proxy.


When starting Docker Desktop with these values the GUI appears on the settings screen but the wait spinner in the middle of the screen spins forever.  The docker status at the bottom of the screen is stuck at yellow with Updating next to it.  The status message in Windows says docker is running and it responds to commands like docker pull.  The Window also does not have my user id and wants me to sign in.

Removing the entries with the leading dot makes the GUI work properly.  This, in turn, will cause problems with the tools that need the leading dot.  In Docker < 2.2 the leading dot was not a problem.  The leading dot entries seems to be preventing the GUI from communicating with the Docker VM.

### Information
  - Is it reproducible?  Yes
  - Is the problem new? Yes, this didn't happen on Docker Desktop 2.1
  - Did the problem appear with an update? Yes.
  - Windows Version: Windows 10 Enterprise 1803
  - Docker Desktop Version: 2.2.0.0
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: No
  - I can't run the diagnostics when the screen in spinning the way it is.

docker version
```console
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea
 Built:             Wed Nov 13 07:22:37 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea
  Built:            Wed Nov 13 07:29:19 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

```
docker info
```console
Client:                                                                                                      
 Debug Mode: false                                                                                           
                                                                                                             
Server:                                                                                                      
 Containers: 20                                                                                              
  Running: 18                                                                                                
  Paused: 0                                                                                                  
  Stopped: 2                                                                                                 
 Images: 12                                                                                                  
 Server Version: 19.03.5                                                                                     
 Storage Driver: overlay2                                                                                    
  Backing Filesystem: extfs                                                                                  
  Supports d_type: true                                                                                      
  Native Overlay Diff: true                                                                                  
 Logging Driver: json-file                                                                                   
 Cgroup Driver: cgroupfs                                                                                     
 Plugins:                                                                                                    
  Volume: local                                                                                              
  Network: bridge host ipvlan macvlan null overlay                                                           
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog                        
 Swarm: inactive                                                                                             
 Runtimes: runc                                                                                              
 Default Runtime: runc                                                                                       
 Init Binary: docker-init                                                                                    
 containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339                                                
 runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657                                                      
 init version: fec3683                                                                                       
 Security Options:                                                                                           
  seccomp                                                                                                    
   Profile: default                                                                                          
 Kernel Version: 4.19.76-linuxkit                                                                            
 Operating System: Docker Desktop                                                                            
 OSType: linux                                                                                               
 Architecture: x86_64                                                                                        
 CPUs: 2                                                                                                     
 Total Memory: 3.824GiB                                                                                      
 Name: docker-desktop                                                                                        
 ID: UR3X:5O7H:RAR6:5B6T:VMSV:6SKU:WD7E:RJNE:Z2NM:QMMA:6SGU:BDK7                                             
 Docker Root Dir: /var/lib/docker                                                                            
 Debug Mode: true                                                                                            
  File Descriptors: 131                                                                                      
  Goroutines: 126                                                                                            
  System Time: 2020-01-22T19:18:27.389459471Z                                                                
  EventsListeners: 3                                                                                         
 HTTP Proxy: http://proxy-server-redacted
 HTTPS Proxy: http://proxy-server-redacted
 No Proxy: .internal,internal,.local,local,.network-redacted.net,network-redacted.net,localhost
 Registry: https://index.docker.io/v1/                                                                       
 Labels:                                                                                                     
 Experimental: false                                                                                         
 Insecure Registries:                                                                                        
  127.0.0.0/8                                                                                                
 Live Restore Enabled: false                                                                                 
 Product License: Community Engine                                                                           
```
### Steps to reproduce the behavior
  1. Add an entry to no_proxy with a leading dot.
  2. Start Docker Desktop
  3. Watch it spin forever when opening Settings.
  4. Shut Docker Desktop down
  5. Remove the leading dot entry
  6. Start Docker Desktop
  7. Open settings and things work fine.

**Comment by mat007** [comment]
Thanks for your report!

We were unable to reproduce the issue, just to be sure here are the steps we’ve tried (we don’t have a proxy):
1. add an environment variable `no_proxy` with the value `.my-network.net,my-network.net`
2. start Docker Desktop (linux containers, not WSL 2) and wait for it to be running
3. open the Settings

Is that correct?

**Comment by cyclump** [comment]
Thank you for your quick response!  

I think you're almost there, but I fear the no_proxy variable is being ignored without the http_proxy and https_proxy variables set.  I have been able to set up the condition for this issue without my corporate proxy.  This example shows the same behavior using a pyhon proxy.  The proxy is at https://github.com/abhinavsingh/proxy.py.

The commands 
```bash
pip install --upgrade proxy.py
proxy --hostname 0.0.0.0 --port 3128 --log-level d
```
will install and run a simple http proxy.  It listens on all adapters on port 3128.

Next set docker to use these proxy settings.  It uses host.docker.internal to hit the adapter ip and not localhost.
- Web Server => `http://host.docker.internal:3128`
- Secure Web Server => `http://host.docker.internal:3128`
- Bypass proxy settings... => `cloud.google.com,gcr.io,.internal,internal,.local,local,amazonaws.com,localhost`

The additional domains preceded with a dot in this setting does not seem to impact anything.

Quit Docker Desktop.  Set your environment variables using this PowerShell script.  It sets the proxy variables in the user's environment.  It puts a dot before amazonaws.com.
```powershell.
[Environment]::SetEnvironmentVariable("http_proxy","http://localhost:3128","User")
[Environment]::SetEnvironmentVariable("https_proxy","http://localhost:3128","User")
[Environment]::SetEnvironmentVariable("no_proxy","cloud.google.com,gcr.io,internal,.amazonaws.com,localhost","User") 
```

Run Docker Desktop and try to access settings.  You'll be greeted with this forever.
![docker-settings-with-dot](https://user-images.githubusercontent.com/7515042/73002714-7430c080-3dd2-11ea-8767-090fe980b365.png)

Quit Docker Desktop.  Set your environment variables using this PowerShell script.  It sets the proxy variables in the user's environment.  It removes the dot before amazonaws.com.
```powershell.
[Environment]::SetEnvironmentVariable("http_proxy","http://localhost:3128","User")
[Environment]::SetEnvironmentVariable("https_proxy","http://localhost:3128","User")
[Environment]::SetEnvironmentVariable("no_proxy","cloud.google.com,gcr.io,internal,amazonaws.com,localhost","User") 
```

Run Docker Desktop and access Settings.  You get your settings screen back.
![docker-settings-without-dot](https://user-images.githubusercontent.com/7515042/73002987-e0abbf80-3dd2-11ea-90e4-113f03537df3.png)

Hopefully this should reproduce the behavior on your system.  Thank you!

**Comment by mat007** [comment]
Thank you, using your very detailed instructions I was able to reproduce this issue!
We’re looking into it.

**Comment by mat007** [comment]
We have a fix, it will be included in the coming 2.2.0.1 release planned for in about a week or so.
Thanks!

**Comment by bkt0031** [comment]
I am having a very similar problem that I think is related to going to the Internet through a proxy. In our case the proxy is a Check Point firewall running transparently as a proxy. It is still a man-in-the-middle in that any SSL/TLS requests are intercepted by the firewall and the server certificate reply contains the firewall certificate which is validated by the companies CA. I had to downgrade back to 2.1.0.5 as I just could not get the GUI to fully come up in 2.2.0.0.

**Comment by stephen-turner** [comment]
We have a new private build which (we believe) fixes this bug. https://download-stage.docker.com/win/stable/42579/Docker%20Desktop%20Installer.exe. Please try it and let us know how you get on.

**Comment by ArPharazon** [comment]
Thanks @cyclump for that excellent diagnosis! I also had a leading dot in my no_proxy env var which i had not noticed until now.

**Comment by cyclump** [comment]
Looks good so far.

**Comment by stephen-turner** [comment]
We have just released 2.2.0.3, which fixes this issue, so closing this ticket. In case any Edge users are watching this ticket, it will be on the Edge channel tomorrow too.
